node/service/src/lib.rs

```diff
@@ -783,22 +784,24 @@ where
 					Ok((time, parachain_inherent, author, randomness))
 				}
 			};
-			// must be defined here to access client runtime API in this function
-			// type Digests = Option<DigestItem> should work
-			let additional_digests_provider =  |nimbus_id: nimbus_primitives::NimbusId, parent: Hash| -> Option<sp_runtime::generic::DigestItem> {
-				let relay_slot_number: polkadot_primitives::v2::Slot = todo!(); // get using runtime API
-				let relay_storage_root: Hash = todo!(); // get using runtime API
-				let key: session_keys_primitives::VrfId = todo!(); // get using runtime API VrfKeyLookup
-				if let Some(vrf_pre_digest) = moonbeam_vrf::vrf_pre_digest::<Hash>(
+			let current_client = client.clone();
+			let current_keystore = keystore.clone();
+			let additional_digests_provider = move |
+				nimbus_id: nimbus_primitives::NimbusId,
+				parent: Hash
+			| -> Option<sp_runtime::generic::DigestItem> {
+				use sp_api::ProvideRuntimeApi;
+				let api = current_client.runtime_api();
+				let relay_slot_number: polkadot_primitives::v2::Slot = api.get_relay_slot_number();
+				let relay_storage_root: Hash = api.get_relay_storage_root();
+				let key: session_keys_primitives::VrfId = api.vrf_key_lookup(nimbus_id)?;
```

Comment:
@librelois I am getting an error related to the runtime API not knowing these functions:
```rust
   Compiling moonbeam-service v0.24.0 (/Users/4meta5/ong/ps/mb/node/service)
error[E0282]: type annotations needed
   --> node/service/src/lib.rs:795:64
    |
795 |                 let relay_slot_number: polkadot_primitives::v2::Slot = api.get_relay_slot_number();
    |                                                                            ^^^^^^^^^^^^^^^^^^^^^ cannot infer type
    |
    = note: type must be known at this point

error[E0599]: no method named `get_relay_slot_number` found for struct `ApiRef<'_, _>` in the current scope
   --> node/service/src/lib.rs:795:64
    |
795 |                 let relay_slot_number: polkadot_primitives::v2::Slot = api.get_relay_slot_number();
    |                                                                            ^^^^^^^^^^^^^^^^^^^^^ method not found in `ApiRef<'_, _>`

error[E0599]: no method named `get_relay_storage_root` found for struct `ApiRef<'_, _>` in the current scope
   --> node/service/src/lib.rs:796:40
    |
796 |                 let relay_storage_root: Hash = api.get_relay_storage_root();
    |                                                    ^^^^^^^^^^^^^^^^^^^^^^ method not found in `ApiRef<'_, _>`

error[E0599]: no method named `vrf_key_lookup` found for struct `ApiRef<'_, _>` in the current scope
   --> node/service/src/lib.rs:797:51
    |
797 |                 let key: session_keys_primitives::VrfId = api.vrf_key_lookup(nimbus_id)?;
    |                                                               ^^^^^^^^^^^^^^ method not found in `ApiRef<'_, _>`

Some errors have detailed explanations: E0282, E0599.
For more information about an error, try `rustc --explain E0282`.
error: could not compile `moonbeam-service` due to 4 previous errors
```

Even though I already added the trait bound `RuntimeApi::RuntimeApi: VrfApi<Block>` at the top of each function.

Let me know if there is something else that must be in scope. Maybe the RuntimeAPI needs to be generic over the output types so I can specify them in the trait bound?